rust/worker/src/execution/operators/limit.rs

```diff
@@ -0,0 +1,317 @@
+        Ok(log_rank + record_rank - mask_rank)
+    }
+
+    // Skip to the start in log and record segment
+    // The following implementation is based on std implementation
+    async fn skip_to_start(&self, skip: usize) -> Result<(usize, usize), LimitError> {
+        if skip == 0 {
+            return Ok((0, 0));
+        }
+
+        let mut size = self
+            .record_segment
+            .get_current_max_offset_id()
+            .load(atomic::Ordering::Relaxed)
+            .max(self.log_oids.max().unwrap_or(0));
+        if size == 0 {
+            return Ok((0, 0));
+        }
+
+        let mut base = 0;
+        while size > 1 {
+            let half = size / 2;
+            let mid = base + half;
+
+            let cmp = self.joint_rank(mid).await?.cmp(&skip);
+            base = if cmp == Ordering::Greater { base } else { mid };
+            size -= half;
+        }
+
+        Ok((
+            self.log_oids.rank(base) as usize - self.log_oids.contains(base) as usize,
+            self.record_segment
+                .get_offset_id_rank(base)
+                .await
+                .map_err(|_| LimitError::RecordSegmentReaderError)?,
+        ))
+    }
+
+    // Skip to the start in the log and record segment, then scan for the specified number of offset ids
+    async fn skip_and_scan(
+        &self,
+        skip: usize,
+        mut fetch: usize,
+    ) -> Result<RoaringBitmap, LimitError> {
+        let record_count = self
+            .record_segment
+            .count()
+            .await
+            .map_err(|_| LimitError::RecordSegmentReaderError)?;
+        let (mut log_index, mut record_index) = self.skip_to_start(skip).await?;
+        let mut merged_result = Vec::new();
+
+        while fetch > 0 {
+            let log_oid = self.log_oids.select(log_index as u32);
+            let record_oid = (record_index < record_count).then_some(
+                self.record_segment
+                    .get_offset_id_at_index(record_index)
+                    .await
+                    .map_err(|_| LimitError::RecordSegmentReaderError)?,
+            );
+            match (log_oid, record_oid) {
+                (_, Some(oid)) if self.mask.contains(oid) => {
+                    record_index += 1;
+                    continue;
+                }
+                (Some(l), Some(r)) => {
+                    if l < r {
+                        merged_result.push(l);
+                        log_index += 1;
+                    } else {
+                        merged_result.push(r);
+                        record_index += 1;
+                    }
+                }
+                (None, Some(oid)) => {
+                    merged_result.push(oid);
+                    record_index += 1;
+                }
+                (Some(oid), None) => {
+                    merged_result.push(oid);
+                    log_index += 1;
+                }
+                _ => {}
```

Comment:
if `fetch` is sized incorrectly I think this could result in an infinite loop? could we `break` or panic with `unreachable()` here?

Reply:
Since `fetch` is always decremented in the end, this should never result in an infinite loop. I changed `{}` to `break` in a later PR as a minor improvement